File: src/main/java/org/apache/sysml/hops/spoof2/plan/SNodeAggregate.java
package org.apache.sysml.hops.spoof2.plan;


import java.util.List;

import org.apache.sysml.hops.Hop.AggOp;

public class SNodeAggregate extends SNode
{
	//reuse of hop aggregation types
	private final AggOp _type;

	public SNodeAggregate(SNode input, AggOp type) {
		super(input);
		_type = type;
	}
	
	public SNodeAggregate(SNode input, AggOp type, List<String> schema) {
		super(input);
		_type = type;
		_schema.addAll(schema);
	}
	
	public AggOp getOp() {
		return _type;
	}

	@Override
	public boolean isIndexAggregator() {
		return false;
	}

	@Override
	public boolean isIndexPropagator() {
		return true;
	}

	@Override
	public String toString() {
		return "agg("+_type.name().toLowerCase()+")";
	}
}

File: src/main/java/org/apache/sysml/hops/spoof2/plan/SNodeAggregate.kt
package org.apache.sysml.hops.spoof2.plan

import org.apache.sysml.hops.Hop.AggOp

/**
 * Aggregate a single input.
 */
class SNodeAggregate(
        val op: AggOp,
        input: SNode,
        aggs: Schema
) : SNode(listOf(input)) {
    var input: SNode
        get() = inputs[0]
        set(v) { inputs[0] = v }

    constructor(op: AggOp, input: SNode, vararg names: AB) :
            this(op, input, input.schema.filterKeys { it in names })
    constructor(op: AggOp, input: SNode, names: Iterable<AB>) :
            this(op, input, input.schema.filterKeys { it in names })

    override fun shallowCopy(newInputs: List<SNode>) = SNodeAggregate(op, newInputs[0], aggs)
    override fun compare(o: SNode) =
            o is SNodeAggregate && o.op == this.op && o.aggs == this.aggs && o.input == this.input

    val aggs: Schema = Schema(aggs)
    init {
        refreshSchema()
    }

    override fun toString() = "agg(${op.name.toLowerCase()}$aggs)"

    override fun checkArity() {
        this.check( inputs.size == 1 ) {"SNodeAggregate should have 1 input but has ${inputs.size}"}
    }

    override fun refreshSchema() {
        // input names minus aggregated names
        schema.setTo(inputs[0].schema)
        schema -= aggs
    }

    fun aggsNotInInputSchema(): Schema {
        return aggs.filterKeys { n -> n !in input.schema }
    }
}
